### PR TITLE
Write Crop Api documentation.

### PR DESCRIPTION
**Pull request by woprrr**
This is a start of doccumentation of CropAPI what do you think @slashrsm ?

Excuse my bad english at time ... if you can correct it, if you found faults.

Original demand : https://www.drupal.org/node/2382637

### CODE REVIEW

File: modules/crop/Implementation_examples.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Implementation Example
+
+Create
+------
+
+This code show an example to create an crop entity and set all positions.
```
Comment (slashrsm) [inline modules/crop/Implementation_examples.md]: This code shows how to create a crop entity...

File: modules/crop/Implementation_examples.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// Save crop with previous values.
+/** @var \Drupal\crop\CropInterface $crop */
+$crop = \Drupal::service->get('entity.manager')->getStorage('crop')->create($values);
+$crop->save();
+```
+
+Update
+------
+
+As CropAPI is an entity it is possible to modify the property taken entity via methods "->set()" & "->save()" for final registration. It is therefore quite possible to update a crop after having loaded a crop with "->loadByProperties()" as the example below.
```
Comment (slashrsm) [inline modules/crop/Implementation_examples.md]: Since Crop is an entity it is possible to modify it's properties using "->set()" & "->save()"  methods.

I don't think second sentence is needed as we're not documenting entity api here.

File: modules/crop/Implementation_examples.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                            'image_style' => ImageStyleName,
+                        ]);
+                        
+$crop->set($field_key, $value, TRUE)->save();
+```
+
+Delete
+------
+
+From the same manner as for the case "Update" you can delete the entity.
```
Comment (slashrsm) [inline modules/crop/Implementation_examples.md]: Crop can be deleted using standard Entity API function.

File: modules/crop/architecture.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# Architecture
+
+Diagram shows basic architecture of Crop API.
+
+![Crop API architecture](images/architecture.png)
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Entity type is called "Crop", not "Crop API".

File: modules/crop/architecture.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+## Components
+
+Entity
+------
+- **Crop** Is a content entity that stores various information related to a crop area. It serves to store and returned all informations corresponding to crop area.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Is a content entity that stores various informations related to the crop area.

File: modules/crop/architecture.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+- **CropType** It has a kind bunlde "CropType" who will inherit all the characteristics of the entity mother "Crop" but also allow a variation of Crop entity and will add the information management bind to a type of crop in particular.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: This is not entirely true. Crop type is bundle for Crop entity. It is as simple as that.

File: modules/crop/architecture.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+Plugin
+------
+- **CropEffect** Allows to apply a crop on an image via a ImageStyle effect.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Crop effect is an image style effect.

File: modules/crop/architecture.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+This diagram shows basic application that allows the effect "Manual Crop".
```
Comment (slashrsm) [inline modules/crop/architecture.md]: What are we trying to say with this diagram?

File: modules/crop/architecture.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+![Manual Crop Effect](images/architecture_of_manual_crop.png)
+
+The effect "Manual Crop Effect" recover the crop has an image associated with two parameters.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Which parameters?

File: modules/crop/architecture.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+- **"uri"** This is an important propertie because it is the only contextual information available when you apply the effect "Manual Crop".
```
Comment (slashrsm) [inline modules/crop/architecture.md]: This is an important property because it is the only contextual information available when you apply the effect to the image.

File: modules/crop/architecture.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+- **"crop_type"** As it is possible to link CROP entity with same image, this is the only parameter that can distinguish several crop_type on the same image.
```
Comment (slashrsm) [inline modules/crop/architecture.md:19]: If there are more crops that need to be attached to an image we need to make sure every one of them uses different Crop type. 

Or something like that...

File: modules/crop/architecture.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+Limitation :
+Currently Crop Api is not able to handle the case where one has several identical crop (uri & crop_type), in this case it returns the first entity found instead of list of corresponding entity.
+@see [Crop API support multiple issue]
```
Comment (slashrsm) [inline modules/crop/architecture.md]: See [Crop API support multiple issue](https://www.drupal.org/node/2617818) for more informations.

File: modules/crop/architecture.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+![Manual Crop Effect](images/architecture_of_manual_crop.png)
+
+The effect "Manual Crop Effect" recover the crop has an image associated with two parameters.
+
+- **"uri"** This is an important propertie because it is the only contextual information available when you apply the effect "Manual Crop".
+- **"crop_type"** As it is possible to link CROP entity with same image, this is the only parameter that can distinguish several crop_type on the same image.
+
+Limitation :
+Currently Crop Api is not able to handle the case where one has several identical crop (uri & crop_type), in this case it returns the first entity found instead of list of corresponding entity.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Due to limitations in Drupal core we're currently not not able to handle the case where one image would have several crops of the same crop type. The oldest crop will always be used in such case.

File: modules/crop/architecture.md [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+- **entity_id** This is the "FID" of "file" corresponding to the image that we want crop.
+- **entity_type** This is the type of crop on which the entity is linked.
+- **uri** This is an important propertie because it is the only contextual information available when you apply the effect "Manual Crop".
+- **x** The position of the top left point on the x-axis.
+- **y** Since x is the position of the top left point on the y-axis.
+- **width** Is the size in width of the zone is copied to the image ("URI").
+- **height** Is the size in height of the area copied image ("URI").
+- **image_style** This is the machine name of the related ImageStyle, it is an optional parameter because nothing really checks at the recovery of the crop with this parameter.
+
+@see Implementation examples to understand more implementation of Crop API.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Where can I fund implementation examples?

File: modules/crop/architecture.md [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+Limitation :
+Currently Crop Api is not able to handle the case where one has several identical crop (uri & crop_type), in this case it returns the first entity found instead of list of corresponding entity.
+@see [Crop API support multiple issue]
+
+[Crop API support multiple issue]: https://www.drupal.org/node/2617818
+
+### Focus on Crop Entity properties
+
+When creating an CROP entity must send a series of properties that will be used for the purposes of the crop area by imageStyle or for the recovery data link to this entity.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Replace 1., 2. and 3. paragraph with something like this:

Properties of crop entity are explained:

File: modules/crop/architecture.md [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+It is important to identify the different elements before Crop API needs to properly render the crop area with good dimmensions.
+
+We will each properties listed on entity Crop :
+
+- **type** This is the machine name of the corresponding crop type use by this entity, it works exactly as "Node" and "NodeType".
+- **entity_id** This is the "FID" of "file" corresponding to the image that we want crop.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: ID of the entity crop relates to. "fid" when entity is a file.

File: modules/crop/architecture.md [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+- **entity_type** This is the type of crop on which the entity is linked.
+- **uri** This is an important propertie because it is the only contextual information available when you apply the effect "Manual Crop".
```
Comment (slashrsm) [inline modules/crop/architecture.md]: URI of the image that is cropped.

File: modules/crop/architecture.md [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+- **x** The position of the top left point on the x-axis.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Position of top left corner of the crop area on x-axis.

File: modules/crop/architecture.md [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+- **y** Since x is the position of the top left point on the y-axis.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Position of top left corner of the crop area on y-axis.

File: modules/crop/architecture.md [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+- **width** Is the size in width of the zone is copied to the image ("URI").
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Width of the crop area.

File: modules/crop/architecture.md [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+- **height** Is the size in height of the area copied image ("URI").
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Height of the crop area.

File: modules/crop/architecture.md [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+- **image_style** This is the machine name of the related ImageStyle, it is an optional parameter because nothing really checks at the recovery of the crop with this parameter.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: We should probably remove this property anyway...

**Comment by woprrr** [comment]
@slashrsm Thanks a lot for you few return !! Sorry for my english again :( I hope this help you a little ... 

This is new version of Doc for u what do you think :) ?

File: modules/crop/architecture.md [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+![Crop architecture](images/architecture.png)
+
+## Components
+
+### Entity
+- **Crop** Is a content entity that stores various informations related to the crop area.
+- **CropType** CropType is bundle for Crop entity.
+
+###Plugin
+- **CropEffect** Crop effect is an image style effect. He apply your crop area into image with information of Crop entity.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: It applies crop to the image using information from the crop entity.

File: modules/crop/architecture.md [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+
+This diagram shows how specifically Crop Image Effect (Manual crop) get informations of corresponding image ressource and apply different effect.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Again. What are we trying to explain with this diagram? I don't think we need it. Sentence above explains that image effect crops image based on the crop entity. That should be enough.

File: modules/crop/architecture.md [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+
+The effect "Manual Crop Effect" recover the crop has an image associated with two parameters "uri" & "crop_type".
+
+- **"uri"** This is an important property because it is the only contextual information available when you apply the effect (manual crop) to the image.
+- **"crop_type"** As it is possible to link CROP entity with same image, this is the only parameter that can distinguish several crop_type on the same image.
+
+Limitation :
+Due to limitations in Drupal core we're currently not not able to handle the case where one image would have several crops of the same crop type. The oldest crop will always be used in such case.
+
+See [Crop API support multiple issue] for more information.
```
Comment (slashrsm) [inline modules/crop/architecture.md:24]: Make this a link and remove sentence below.

File: modules/crop/architecture.md [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+
+[Crop API support multiple issue]: https://www.drupal.org/node/2617818
+
+## Focus on properties of crop entity
+
+When creating an CROP entity must send a series of properties that will be used for the purposes of the crop area by imageStyle or for the recovery data link to this entity.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: Remove everything from line 34 to line 37 and replace it with:

"Base field of crop entity explained:"

File: modules/crop/architecture.md [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+
+- **type** This is the machine name of the corresponding crop type use by this entity, it works exactly as "Node" and "NodeType".
+- **entity_id** ID of the entity crop relates to. "fid" when entity is a file.
+- **entity_type** This is the type of crop on which the entity is linked.
+- **uri** URI of the image that is cropped.
+- **x** Position of top left corner of the crop area on x-axis.
+- **y** Position of top left corner of the crop area on y-axis.
+- **width** Width of the crop area.
+- **height** Height of the crop area.
+- **image_style** This is the machine name of the related ImageStyle, it is an optional parameter because nothing really checks at the recovery of the crop with this parameter.
```
Comment (slashrsm) [inline modules/crop/architecture.md]: As said before. Remove this. We should remove it from crop entity also.

File: modules/entity_browser/intro.md [hunk 28]
```diff
@@ -1,5 +1,6 @@
 # Entity browser
 
+The Crop api provides storage for all cropping solution. Is able to handle crops based on region and focal point concepts. Also comes with some very basic image effects.
```
Comment (slashrsm) [inline modules/entity_browser/intro.md]: How is this related to entity browser?

Comment (woprrr) [inline modules/entity_browser/intro.md]: An error :o !! I remove this i ve don't see sorry

File: modules/crop/architecture.md (hunk 26 above)
Comment (woprrr) [inline modules/crop/architecture.md]: I thought it would be necessary to detail a bit like an intro before the explanation

File: modules/crop/architecture.md (hunk 24 above)
Comment (woprrr) [inline modules/crop/architecture.md]: I think it would be good to clarify how Crop entity is used by imageStyle and how the recovery of crop has applied to the resource. It seems to me the scope Crop entity. I have not made voluntarily detail the portion not off topic.

File: modules/crop/architecture.md (hunk 25 above)
Comment (woprrr) [inline modules/crop/architecture.md:24]: The line 28 is already linked to issue. What you need remove ?

File: modules/crop/architecture.md (hunk 27 above)
Comment (woprrr) [inline modules/crop/architecture.md]: Because we are based on file URI, but it might be good to keep to the API that uses crop entity. To be on the retrieve the corresponding cropType wanted imageStyle

**Comment by slashrsm** [comment]
Merged #12 that was based on top of this.

**Comment by slashrsm** [comment]
Thank you both!
